fuse.py

```diff
@@ -616,11 +633,42 @@ def _wrapper(func, *args, **kwargs):
         'Decorator for the methods that follow'
 
         try:
-            return func(*args, **kwargs) or 0
-        except OSError as e:
-            return -(e.errno or errno.EFAULT)
-        except:
-            print_exc()
+            if func.__name__ == "init":
+                # init may not fail, as its return code is just stored as
+                # private_data field of struct fuse_context
+                return func(*args, **kwargs) or 0
+
+            else:
+                try:
+                    return func(*args, **kwargs) or 0
+
+                except OSError as e:
+                    if e.errno > 0:
+                        log.debug(
+                            "FUSE operation %s raised a %s, returning errno %s.",
+                            func.__name__, type(e), e.errno, exc_info=True)
+                        return -e.errno
```

Comment:
@N-Coder It seems that this takes a positive errno and then negates it. In the PR description, it said this was support to _not_ forward negative errors. Is this a mistake or am I misunderstanding what is happening  in the code?

Reply:
Jep, this is correct and actually the part that was also already there before in `return -(e.errno or errno.EFAULT)`, where `e.errno` was simply assumed to be positive (which is not always true). This is due to standard errnos being [positive](http://www.virtsync.com/c-error-codes-include-errno) and FUSE interpreting [negative](http://libfuse.github.io/doxygen/structfuse__operations.html) values as the respective error (while a positive value would indicate success and e.g. represent a file handle):
>  A major exception is that instead of returning an error in 'errno' [the global variable], the operation should return the negated error value (-errno) directly.

This works fine for most OSErrors carrying standard error codes...  [the following is mostly copied from my original issue description]
> But, some subclasses of OSError (like [herror and gaierror](https://docs.python.org/3/library/socket.html#socket.herror)) use errnos for values that don't correspond to the normal / standard Linux OS errnos (but to host name resolution or getaddrinfo error codes).
> On my system, the errnos for gaierror are:
> ```
> {'EAI_ADDRFAMILY': -9, 'EAI_AGAIN': -3, 'EAI_BADFLAGS': -1, 'EAI_FAIL': -4, 'EAI_FAMILY': -6, 'EAI_MEMORY': -10, 'EAI_NODATA': -5, 'EAI_NONAME': -2, 'EAI_OVERFLOW': -12, 'EAI_SERVICE': -8, 'EAI_SOCKTYPE': -7, 'EAI_SYSTEM': -11}
> ```
> If such an error code (note the negative sign for the EAI_* errors, while the normal OS errnos have a positive sign) is returned to libfuse, a warning about an invalid error code might be issued, or the library might interpret this now positive value as valid file handle, leading to a lot of problems.
> So those exception should be translated to standard errnos, e.g. EHOSTUNREACH, instead.

As we currently have no translation between specific errors (like herror and gaierror) and I don't know whether there are other places where non-standard errnos are used, I just went with returning a generic errno.EINVAL.

Reply:
Okay, this makes a bit more sense to me now. Everything else looks good to me so I will merge now.